spec/services/pii/nist_encryption_spec.rb

```diff
@@ -0,0 +1,152 @@
+
+      # Z1, Z2 = scrypt(S, password)   # split 256-bit output into two halves
+      hashed_password = PasswordHasher.hash(password, salt)
+
+      expect(hex_to_bin(hashed_password.z1).length).to eq 16
+      expect(hex_to_bin(hashed_password.z2).length).to eq 16
+
+      expect do
+        SCrypt::Password.new(hashed_password.as_scrypt_hash)
+      end.to_not raise_error
+    end
+  end
+
+  describe 'KMS encryption of random R' do
+    it 'creates encrypted key D and hash E' do
+      random_R, ciphered_R = stub_aws_kms_client
+      allow(Pii::Cipher).to receive(:random_key).and_return(random_R)
+
+      password = 'a long sekrit'
+      salt = SecureRandom.random_bytes(32)
+      hashed_password = PasswordHasher.hash(password, salt)
+
+      # D = KMS_GCM_Encrypt(key=server_secret, plaintext=R) ^ Z1
+      # E = hash( Z2 + R )
+      encrypted_D = hashed_password.xor(ciphered_R)
+      hash_E = OpenSSL::Digest::SHA256.hexdigest(hashed_password.z2 + random_R)
+
+      key_maker = Pii::KeyMaker.new
+      key_maker.make(hashed_password)
+
+      expect(hashed_password.encrypted_d).to eq encrypted_D
+      expect(hashed_password.hash_e).to eq hash_E
+    end
+  end
+
+  describe 'password storage and login' do
+    it 'creates hash F of Z2 using auto-generated salt' do
+      random_R, ciphered_R = stub_aws_kms_client
+      allow(Pii::Cipher).to receive(:random_key).and_return(random_R)
+
+      password = 'a long sekrit'
+      user = create(:user, password: password)
+
+      expect(user.valid_password?(password)).to eq true
+      expect(user.hashed_password).to be_a HashedPassword
+      expect(user.hashed_password.random_r).to eq random_R
+      expect(user.encryption_key).to_not be_nil
+      expect(user.password_salt).to_not be_nil
+
+      hash_E = OpenSSL::Digest::SHA256.hexdigest(user.hashed_password.z2 + random_R)
+      hash_F = OpenSSL::Digest::SHA256.hexdigest(hash_E)
+
+      expect(user.encrypted_password).to eq hash_F
+      expect(user.hashed_password.hash_e).to eq hash_E
+      expect(user.hashed_password.hash_f).to eq hash_F
+
+      key_maker = Pii::KeyMaker.new
+      expect(key_maker.unlock(user.hashed_password, user.encryption_key)).to eq hash_E
+
+      encrypted_D = Base64.strict_decode64(user.encryption_key)
+
+      expect(user.hashed_password.xor(ciphered_R)).to eq encrypted_D
+    end
+  end
+
+  describe 'PII encryption' do
+    it 'creates encrypted payload C with KMS-encrypted key D using local AES cipher' do
+      random_R, ciphered_R = stub_aws_kms_client
+      allow(Pii::Cipher).to receive(:random_key).and_return(random_R)
+
+      password = 'a long sekrit'
+      salt = SecureRandom.random_bytes(32)
+      hashed_password = PasswordHasher.hash(password, salt)
+      pii = 'some sensitive stuff'
+
+      # D = KMS_GCM_Encrypt(key=server_secret, plaintext=R) ^ Z1
+      # E = hash( Z2 + R )
+      # C = GCM_Encrypt(key = E, plaintext=PII)  # occurs outside AWS-KMS
+      # Store C and D.
```

Comment:
`spec/services/pii/nist_encryption_spec.rb` is intentionally verbose and overly commented. The intended audience is NIST and other crypto folks, so that they can see, as closely as possible, the correlation between the model documentation and the implementation.